frontstage/controllers/oauth_controller.py

```diff
@@ -0,0 +1,44 @@
+import json
+import logging
+
+from flask import current_app as app
+import requests
+from structlog import wrap_logger
+
+from frontstage.exceptions.exceptions import ApiError, OAuth2Error
+
+
+logger = wrap_logger(logging.getLogger(__name__))
+
+
+def sign_in(username, password):
+    logger.debug('Attempting to retrieve OAuth2 token for sign-in')
+    
+    url = f"{app.config['OAUTH_SERVICE_URL']}/api/v1/tokens/"
+    data = {
+        'grant_type': 'password',
+        'client_id': app.config['OAUTH_CLIENT_ID'],
+        'client_secret': app.config['OAUTH_CLIENT_SECRET'],
+        'username': username,
+        'password': password,
+    }
+    headers = {
+        'Accept': 'application/json',
+        'Content-Type': 'application/x-www-form-urlencoded;charset=UTF-8',
+    }
+    response = requests.post(url, headers=headers, auth=app.config['OAUTH_BASIC_AUTH'], data=data)
+
+    try:
+        response.raise_for_status()
+    except requests.exceptions.HTTPError:
+        if response.status_code == 401:
+            oauth2_error = json.loads(response.text).get('error').get('data', {}).get('detail')
+            logger.exception('Authentication error in oauth2 service', error=oauth2_error)
+            raise OAuth2Error(response, message=oauth2_error)
+        else:
+            logger.exception('Failed to retrieve OAuth2 token')
+        raise ApiError(response)
+
+    oauth2_token = json.loads(response.text)
```

Comment:
Decode errors should be handled here.

Reply:
Unexpected errors are handled by the generic [catch all](https://github.com/ONSdigital/ras-frontstage/blob/master/frontstage/error_handlers.py#L45). I'm not sure whether there's an intention to change to a more defensive style but probably worth raising.